ksp_plugin/orbit_analyser.cpp

```diff
@@ -0,0 +1,212 @@
+﻿
+#include "ksp_plugin/orbit_analyser.hpp"
+
+#include "physics/body_centred_non_rotating_dynamic_frame.hpp"
+#include "physics/discrete_trajectory.hpp"
+
+namespace principia {
+namespace ksp_plugin {
+namespace internal_orbit_analyser {
+
+using geometry::Frame;
+using physics::DiscreteTrajectory;
+using physics::MasslessBody;
+using physics::BodyCentredNonRotatingDynamicFrame;
+using quantities::IsFinite;
+
+OrbitAnalyser::OrbitAnalyser(not_null<Ephemeris<Barycentric>*> const ephemeris,
+                             Ephemeris<Barycentric>::FixedStepParameters const
+                                 analysed_trajectory_parameters)
+    : ephemeris_(ephemeris),
+      analysed_trajectory_parameters_(analysed_trajectory_parameters),
+      analyser_([this] { RepeatedlyAnalyseOrbit(); }) {}
+
+OrbitAnalyser::~OrbitAnalyser() {
+  if (analyser_.joinable()) {
+    keep_analysing_ = false;
+    analyser_.join();
+  }
+}
+
+void OrbitAnalyser::RequestAnalysis(
+    Instant const& first_time,
+    DegreesOfFreedom<Barycentric> const& first_degrees_of_freedom,
+    Time const& mission_duration,
+    not_null<RotatingBody<Barycentric> const*> primary) {
+  Ephemeris<Barycentric>::Guard guard(ephemeris_);
+  if (ephemeris_->t_min() > first_time) {
+    // Too much has been forgotten; we cannot perform this analysis.
+    return;
+  }
+  absl::MutexLock l(&lock_);
+  parameters_ = {std::move(guard),
+                 first_time,
+                 first_degrees_of_freedom,
+                 mission_duration,
+                 primary};
+}
+
+void OrbitAnalyser::RefreshAnalysis() {
+  absl::MutexLock l(&lock_);
+  if (next_analysis_.has_value()) {
+    analysis_ = next_analysis_;
+    next_analysis_.reset();
+  }
+}
+
+OrbitAnalyser::Analysis* OrbitAnalyser::analysis() {
+  return analysis_.has_value() ? &*analysis_ : nullptr;
+}
+
+int8_t OrbitAnalyser::next_analysis_percentage() const {
+  return next_analysis_percentage_;
+}
+
+void OrbitAnalyser::RepeatedlyAnalyseOrbit() {
+  while (keep_analysing_) {
```

Comment:
Preference for `for (;;)` with exits at a few critical places.  As written it's a bit hard to be sure that we exit after/before every slow operation.